plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt

```diff
@@ -231,47 +236,82 @@ data class ClassTypeEmbedding(val className: ScopedKotlinName, val isInterface:
-            .filter { it.accessPolicy == AccessPolicy.ALWAYS_READABLE }
-            .mapNotNull { it.type.predicateAccessInvariant()?.fillHole(PrimitiveFieldAccess(subjectEmbedding, it)) }
+        val accessFieldPredicates = filteredFields
+            .mapNotNull { it.type.sharedPredicateAccessInvariant()?.fillHole(PrimitiveFieldAccess(subjectEmbedding, it)) }
 
-        val fieldsProvenInvariants = fields.values
-            .filter { it.accessPolicy == AccessPolicy.ALWAYS_READABLE }
+        val fieldsProvenInvariants = filteredFields
             .flatMap { it.type.provenInvariants().fillHoles(PrimitiveFieldAccess(subjectEmbedding, it)) }
 
-        val accessSuperTypesPredicates = superTypes
-            .filterIsInstance<ClassTypeEmbedding>()
+        val accessSuperTypesPredicates = classSuperTypes
             .map { PredicateAccessTypeInvariantEmbedding(it.name, PermExp.WildcardPerm()).fillHole(subjectEmbedding) }
 
         val body = (accessFields + accessFieldPredicates + accessSuperTypesPredicates + fieldsProvenInvariants).toConjunction()
         return Predicate(name, listOf(subjectEmbedding.toLocalVarDecl()), body.pureToViper(toBuiltin = true))
     }
 
+    private fun initUniquePredicate(): Predicate {
+        val filteredFields = fields.values.filterIsInstance<UserFieldEmbedding>()
+
+        val accessFields = filteredFields.map {
+            if (it.symbol.isVal) {
+                FieldAccessTypeInvariantEmbedding(it, PermExp.WildcardPerm()).fillHole(subjectEmbedding)
+            } else {
+                FieldAccessTypeInvariantEmbedding(it, PermExp.FullPerm()).fillHole(subjectEmbedding)
+            }
+        }
+
+        val accessFieldSharedPredicates = filteredFields
+            .mapNotNull { it.type.sharedPredicateAccessInvariant()?.fillHole(PrimitiveFieldAccess(subjectEmbedding, it)) }
+
+        val accessFieldUniquePredicates = filteredFields
+            .filter { it.isUnique }
+            .mapNotNull { it.type.uniquePredicateAccessInvariant()?.fillHole(PrimitiveFieldAccess(subjectEmbedding, it)) }
+
+        val fieldsProvenInvariants = filteredFields
+            .flatMap { it.type.provenInvariants().fillHoles(PrimitiveFieldAccess(subjectEmbedding, it)) }
+
+        val accessSuperTypesPredicates = classSuperTypes
+            .map { PredicateAccessTypeInvariantEmbedding(it.uniquePredicateName, PermExp.FullPerm()).fillHole(subjectEmbedding) }
+
+        val body =
+            (accessFields + accessFieldSharedPredicates + accessFieldUniquePredicates + accessSuperTypesPredicates + fieldsProvenInvariants).toConjunction()
+        return Predicate(uniquePredicateName, listOf(subjectEmbedding.toLocalVarDecl()), body.pureToViper(toBuiltin = true))
+    }
+
     // TODO: incorporate generic parameters.
     override val name = object : MangledName {
         override val mangled: String = "T_class_${className.mangled}"
     }
 
+    private val uniquePredicateName = object : MangledName {
```

Comment:
I think we can also mangle the name of the shared predicate in the same way. Since this would modify a lot of tests, probably it's better to do it in a separate PR.